File: zig/timer-ncurses/src/sTime.zig
const std_time = @import("std").time;

pub const sTime = struct {
    timestamp: i64,
    tzdiff: ?i16,

    pub fn update(self: *sTime, tzdiff: ?i8) void {
        if (self.tzdiff == null and tzdiff != null) {
            self.tzdiff = @as(i16, tzdiff.?);
            return self.update(null);
        } else if (self.tzdiff != null) {
            self.timestamp = std_time.timestamp() + self.tzdiff.? * std_time.s_per_hour;
            return;
        }
        self.timestamp = std_time.timestamp();
        return;
    }

    pub fn utcnow() sTime {
        return sTime{ .timestamp = std_time.timestamp() };
    }
    pub fn now(tzdiff: i8) sTime {
        return sTime{ .timestamp = std_time.timestamp() + @as(i16, tzdiff) * std_time.s_per_hour, .tzdiff = tzdiff };
    }

    pub fn from_timestamp(timestamp: i64, tzdiff: ?i16) sTime {
        return sTime{ .timestamp = timestamp, .tzdiff = tzdiff };
    }

    pub fn from_mins(mins: i64, tzdiff: ?i16) sTime {
        return sTime{ .timestamp = mins * std_time.s_per_min, .tzdiff = tzdiff };
    }

    pub fn hours(self: *sTime) i64 {
        // integer part of division is the hours
        return @mod(@divTrunc(self.timestamp, std_time.s_per_hour), 24);
    }

    pub fn minutes(self: *sTime) i64 {
        return @mod(@divTrunc(self.timestamp, std_time.s_per_min), 60);
    }

    pub fn seconds(self: *sTime) i64 {
        return @mod(self.timestamp, 60);
    }
};

File: rust/ratatui_timer/src/s_time.rs
use std::time::{SystemTime, UNIX_EPOCH};

pub struct STime {
    pub timestamp: i64,
    tzdiff: Option<i16>,
}

impl STime {
    pub fn update(&mut self, tzdiff: Option<i8>) {
        if let Some(tzdiff) = tzdiff {
            self.tzdiff = Some(tzdiff as i16);
            self.update(None);
        } else if let Some(tzdiff) = self.tzdiff {
            self.timestamp = Self::now(tzdiff as i8).timestamp;
        } else {
            self.timestamp = Self::utcnow().timestamp;
        }
    }

    pub fn utcnow() -> Self {
        Self {
            timestamp: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .expect("Failed to get system time")
                .as_secs() as i64,
            tzdiff: None,
        }
    }

    pub fn now(tzdiff: i8) -> Self {
        Self {
            timestamp: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .expect("Failed to get system time")
                .as_secs() as i64
                + tzdiff as i64 * 3600,
            tzdiff: Some(tzdiff as i16),
        }
    }

    pub fn from_timestamp(timestamp: i64, tzdiff: Option<i16>) -> Self {
        Self { timestamp, tzdiff }
    }

    pub fn from_mins(mins: i64, tzdiff: Option<i16>) -> Self {
        Self {
            timestamp: mins * 60,
            tzdiff,
        }
    }

    pub fn hours(&self) -> i64 {
        (self.timestamp / 3600) % 24
    }

    pub fn minutes(&self) -> i64 {
        (self.timestamp / 60) % 60
    }

    pub fn seconds(&self) -> i64 {
        self.timestamp % 60
    }
}
